src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs

```diff
@@ -32,58 +37,70 @@ public class DefaultAuthenticator : IAuthenticator
         public DefaultAuthenticator(UsernamePasswordCredential credential)
             : this(credential, new DefaultRandomStringGenerator())
         {
         }
 
-        internal DefaultAuthenticator(UsernamePasswordCredential credential, IRandomStringGenerator randomStringGenerator)
+        internal DefaultAuthenticator(UsernamePasswordCredential credential,
+            IRandomStringGenerator randomStringGenerator)
         {
             _credential = Ensure.IsNotNull(credential, nameof(credential));
             _randomStringGenerator = Ensure.IsNotNull(randomStringGenerator, nameof(randomStringGenerator));
         }
 
         // properties
-        /// <inheritdoc/>
-        public string Name
-        {
-            get { return "DEFAULT"; }
-        }
+        /// <inheritdoc />
+        public string Name => "DEFAULT";
 
         // methods
-        /// <inheritdoc/>
-        public void Authenticate(IConnection connection, ConnectionDescription description, CancellationToken cancellationToken)
+        /// <inheritdoc />
+        public void Authenticate(IConnection connection, ConnectionDescription description,
+            CancellationToken cancellationToken)
         {
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
-            var authenticator = CreateAuthenticator(description);
+            var authenticator = CreateAuthenticator(connection, description);
             authenticator.Authenticate(connection, description, cancellationToken);
         }
 
-        /// <inheritdoc/>
-        public Task AuthenticateAsync(IConnection connection, ConnectionDescription description, CancellationToken cancellationToken)
+        /// <inheritdoc />
+        public async Task AuthenticateAsync(IConnection connection, ConnectionDescription description,
+            CancellationToken cancellationToken)
         {
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
-            var authenticator = CreateAuthenticator(description);
-            return authenticator.AuthenticateAsync(connection, description, cancellationToken);
+            var authenticator = CreateAuthenticator(connection, description);
+            await authenticator.AuthenticateAsync(connection, description, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static BsonDocument CreateSaslSupportedMechsRequest(string authenticationDatabaseName, string userName)
+        {
+            return new BsonDocument {{"saslSupportedMechs", $"{authenticationDatabaseName}.{userName}"}};
+        }
+
+
+        private IAuthenticator CreateAuthenticator(IConnection connection, ConnectionDescription description)
+        {
+            var saslSupportedMechs = description.IsMasterResult.SaslSupportedMechs.ToList();
+           
+            return !saslSupportedMechs.Any()
```

Comment:
Implements the following: https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#authentication-handshake

Reply:
Fixed.